CalibPPS/TimingCalibration/plugins/PPSTimingCalibrationPCLWorker.cc

```diff
@@ -0,0 +1,126 @@
+
+class PPSTimingCalibrationPCLWorker : public DQMGlobalEDAnalyzer<TimingCalibrationHistograms> {
+public:
+  explicit PPSTimingCalibrationPCLWorker(const edm::ParameterSet&);
+
+  void dqmAnalyze(const edm::Event&, const edm::EventSetup&, const TimingCalibrationHistograms&) const override;
+
+  static void fillDescriptions(edm::ConfigurationDescriptions&);
+
+private:
+  void bookHistograms(DQMStore::IBooker&,
+                      const edm::Run&,
+                      const edm::EventSetup&,
+                      TimingCalibrationHistograms&) const override;
+
+  edm::EDGetTokenT<edm::DetSetVector<CTPPSDiamondRecHit>> diamondRecHitToken_;
+  edm::ESGetToken<CTPPSGeometry, VeryForwardRealGeometryRecord> geomEsToken_;
+
+  const std::string dqmDir_;
+};
+
+//------------------------------------------------------------------------------
+
+PPSTimingCalibrationPCLWorker::PPSTimingCalibrationPCLWorker(const edm::ParameterSet& iConfig)
+    : diamondRecHitToken_(
+          consumes<edm::DetSetVector<CTPPSDiamondRecHit>>(iConfig.getParameter<edm::InputTag>("diamondRecHitTag"))),
+      geomEsToken_(esConsumes<edm::Transition::BeginRun>()),
+      dqmDir_(iConfig.getParameter<std::string>("dqmDir")) {}
+
+//------------------------------------------------------------------------------
+
+void PPSTimingCalibrationPCLWorker::bookHistograms(DQMStore::IBooker& iBooker,
+                                                   const edm::Run& iRun,
+                                                   const edm::EventSetup& iSetup,
+                                                   TimingCalibrationHistograms& iHists) const {
+  iBooker.cd();
+  iBooker.setCurrentFolder(dqmDir_);
+  std::string ch_name;
+
+  const auto& geom = iSetup.getData(geomEsToken_);
+  for (auto it = geom.beginSensor(); it != geom.endSensor(); ++it) {
+    if (!CTPPSDiamondDetId::check(it->first))
+      continue;
+    const CTPPSDiamondDetId detid(it->first);
+    if (detid.station() != 1)
+      continue;
+    detid.channelName(ch_name);
+    iHists.leadingTime[detid.rawId()] = iBooker.book1D("t_" + ch_name, ch_name + ";t (ns);Entries", 1200, -60., 60.);
+    iHists.toT[detid.rawId()] = iBooker.book1D("tot_" + ch_name, ch_name + ";ToT (ns);Entries", 100, -20., 20.);
+    iHists.leadingTimeVsToT[detid.rawId()] =
+        iBooker.book2D("tvstot_" + ch_name, ch_name + ";ToT (ns);t (ns)", 240, 0., 60., 450, -20., 25.);
+  }
+}
+
+//------------------------------------------------------------------------------
+
+void PPSTimingCalibrationPCLWorker::dqmAnalyze(const edm::Event& iEvent,
+                                               const edm::EventSetup& iSetup,
+                                               const TimingCalibrationHistograms& iHists) const {
+  // then extract the rechits information for later processing
+  edm::Handle<edm::DetSetVector<CTPPSDiamondRecHit>> dsv_rechits;
+  iEvent.getByToken(diamondRecHitToken_, dsv_rechits);
```

Comment:
Now I realise this non-thread safe ES condition retrieval was completely forgotten! Will be updated in a later push.

Reply:
Hi, You mean you will fix this before the merge of this PR or after?

Reply:
Before the merge, likely. I guess more review is expected, so I do not want to bug @cmsbuild with too many "please test"...

Reply:
Ok. No problem.